docs/rfcs/2023-03-29-file-external-table/rfc.md

```diff
@@ -0,0 +1,86 @@
+---
+Feature Name: "File external table"
+Tracking Issue: https://github.com/GreptimeTeam/greptimedb/issues/1041
+Date: 2023-03-08
+Author: "Xu Wenkang <wenymedia@gmail.com>"
+---
+
+File external table
+
+---
+
+# Summary
+
+Allows users to perform SQL queries on files
+
+# Motivation
+
+User data may already exist in other storages, i.g., file systems/s3, etc. in CSV, parquet, JSON format, etc. We can provide users the ability to perform SQL queries on these files.
+
+
+# Details
+## Overview
+![overview](external-table-engine-overview.png)
+
+We implement a file table engine that creates an external table by accepting user-specified file paths and treating all records as immutable.
+
+1. File Format Decoder: decode files to the `RecordBatch` stream.
+2. Table engine: stores `RecordBatch` in memory and provides scan ability.
+
+## Life cycle
+
+### Register table
+1. Creates metadata and stores it in memory.
+2. Creates the table via file table engine.
+3. Registers table to `CatalogProvider` and it won't register table to `SystemCatalog` (a place that will persist all user-created tables).
+4. Registers table to `SessionLayer` (currently not exists).
+
+### Clean up
+For now, GreptimeDB still lacks a session layer. The system can't release an external table automatically when the user session is closed. But it still has serval ways to allow users to release an external table. 
+1. We enable users to remove an external table manually. 
+2. After the database restarts, all external tables will disappear.
+
+#### Release external tables when starts
+The system  won't persist external tables to disk. Therefore, all external tables wouldn't be recovered after the system restarts.
+
+#### Deregister table
+1. Fetches the target table from `CatalogProvider`
+2. Deregisters the target table in `CatalogProvider`
+3. Finds the target table engine
+4. Drops the target table
+
+## Drawbacks
+- Unsupported filter pushdown
```

Comment:
Can't pushdown filter may be a big issue when querying large dataset.

Reply:
> Can't pushdown filter may be a big issue when querying large dataset.

Therefore, it may work well for the a smallaset(~GB CSV), If users mightry to query on a large CSV file. it may be better to convert CSV to `parquet` format, then we can utilize the row groups, page index to pushdown the some filters.

Reply:
> Can't pushdown filter may be a big issue when querying large dataset.

Do we really need to use an external table that requires a large amount of data? Wouldn't it be better to simply copy the data into the database using `COPY FROM`?

Reply:
Well, as far as I understand it, the file external table is more like a "view upon files", while `copy from` deals with actual tables that have all sorts of powers to read, write and alter data.

Reply:
Yep, the user still can import the data as a `normal` table. So i think we can go ahead.